src/util/address.rs

```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
     }
 }
 
+/// Version of the WitnessProgram: first byte of `scriptPubkey` in
+/// transaction output for transactions starting with opcodes ranging from 0
+/// to 16 (inclusive).
+///
+/// Structure helps to limit possible version of the witness according to the
+/// specification; if a plain `u8` type will be used instead it will mean that
+/// version > 16, which is incorrect.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(u8)]
+pub enum WitnessVersion {
+    /// Initial version of Witness Program. Used for P2WPKH and P2WPK outputs
+    V0 = 0,
+    /// Version of Witness Program used for Taproot P2TR outputs
+    V1 = 1,
+    /// Future (unsupported) version of Witness Program
+    V2 = 2,
+    /// Future (unsupported) version of Witness Program
+    V3 = 3,
+    /// Future (unsupported) version of Witness Program
+    V4 = 4,
+    /// Future (unsupported) version of Witness Program
+    V5 = 5,
+    /// Future (unsupported) version of Witness Program
+    V6 = 6,
+    /// Future (unsupported) version of Witness Program
+    V7 = 7,
+    /// Future (unsupported) version of Witness Program
+    V8 = 8,
+    /// Future (unsupported) version of Witness Program
+    V9 = 9,
+    /// Future (unsupported) version of Witness Program
+    V10 = 10,
+    /// Future (unsupported) version of Witness Program
+    V11 = 11,
+    /// Future (unsupported) version of Witness Program
+    V12 = 12,
+    /// Future (unsupported) version of Witness Program
+    V13 = 13,
+    /// Future (unsupported) version of Witness Program
+    V14 = 14,
+    /// Future (unsupported) version of Witness Program
+    V15 = 15,
+    /// Future (unsupported) version of Witness Program
+    V16 = 16,
+}
+
+impl fmt::Display for WitnessVersion {
```

Comment:
Might be nice to add doc explaining that numeric representation is printed without any prefix (like `v`) or suffix.